File: employee-service-java/src/main/java/com/example/employeeservice/controller/EmployeeController.java
package com.example.employeeservice.controller;

import com.example.employeeservice.model.Employee;
import com.example.employeeservice.service.EmployeeService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;
import lombok.extern.slf4j.Slf4j;

@RestController
@RequestMapping("/api/employees")
@Slf4j
public class EmployeeController {

    private final EmployeeService employeeService;
    
    @Autowired
    public EmployeeController(EmployeeService employeeService) {
        this.employeeService = employeeService;
    }
    
    @GetMapping
    public ResponseEntity<List<Employee>> getAllEmployees() {
        log.info("REST request to get all employees");
        List<Employee> employees = employeeService.getAllEmployees();
        return new ResponseEntity<>(employees, HttpStatus.OK);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<Employee> getEmployeeById(@PathVariable Long id) {
        log.info("REST request to get employee with ID: {}", id);
        
        Optional<Employee> employee = employeeService.getEmployeeById(id);
        
        return employee.map(value -> new ResponseEntity<>(value, HttpStatus.OK))
                .orElseGet(() -> {
                    log.error("Employee with ID: {} not found", id);
                    return new ResponseEntity<>(HttpStatus.NOT_FOUND);
                });
    }
    
    @PostMapping
    public ResponseEntity<Employee> createEmployee(@RequestBody Employee employee) {
        log.info("REST request to create employee: {}", employee);
        
        try {
            Employee newEmployee = employeeService.createEmployee(employee);
            return new ResponseEntity<>(newEmployee, HttpStatus.CREATED);
        } catch (Exception e) {
            log.error("Failed to create employee: {}", e.getMessage());
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @PutMapping("/{id}")
    public ResponseEntity<Employee> updateEmployee(@PathVariable Long id, @RequestBody Employee employee) {
        log.info("REST request to update employee with ID: {}", id);
        
        try {
            Employee updatedEmployee = employeeService.updateEmployee(id, employee);
            return new ResponseEntity<>(updatedEmployee, HttpStatus.OK);
        } catch (RuntimeException e) {
            log.error("Failed to update employee: {}", e.getMessage());
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            log.error("Failed to update employee: {}", e.getMessage());
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
    
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteEmployee(@PathVariable Long id) {
        log.info("REST request to delete employee with ID: {}", id);
        
        try {
            employeeService.deleteEmployee(id);
            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
        } catch (RuntimeException e) {
            log.error("Failed to delete employee: {}", e.getMessage());
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        } catch (Exception e) {
            log.error("Failed to delete employee: {}", e.getMessage());
            return new ResponseEntity<>(HttpStatus.INTERNAL_SERVER_ERROR);
        }
    }
}

File: employee-service-scala/src/main/scala/com/example/employeeservice/controller_bak/EmployeeController.scala
package com.example.employeeservice.controller

import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import com.example.employeeservice.model.{EmployeeRequest, ErrorResponse}
import com.example.employeeservice.service.EmployeeService
import com.typesafe.scalalogging.LazyLogging
import de.heikoseeberger.akkahttpcirce.FailFastCirceSupport
import io.circe.generic.auto._

import scala.concurrent.ExecutionContext
import scala.util.{Failure, Success}

class EmployeeController(service: EmployeeService)(implicit ec: ExecutionContext) 
    extends LazyLogging with FailFastCirceSupport {
  
  val routes: Route = {
    pathPrefix("api" / "employees") {
      concat(
        pathEnd {
          concat(
            get {
              logger.debug("GET request to fetch all employees")
              onComplete(service.getAllEmployees()) {
                case Success(employees) => complete(employees)
                case Failure(ex) => 
                  logger.error(s"Error fetching all employees: ${ex.getMessage}", ex)
                  complete(StatusCodes.InternalServerError -> ErrorResponse(s"Failed to get employees: ${ex.getMessage}"))
              }
            },
            post {
              entity(as[EmployeeRequest]) { request =>
                logger.debug(s"POST request to create employee: $request")
                onComplete(service.createEmployee(request)) {
                  case Success(employee) => complete(StatusCodes.Created -> employee)
                  case Failure(ex) => 
                    logger.error(s"Error creating employee: ${ex.getMessage}", ex)
                    complete(StatusCodes.InternalServerError -> ErrorResponse(s"Failed to create employee: ${ex.getMessage}"))
                }
              }
            }
          )
        },
        path(Segment) { id =>
          concat(
            get {
              logger.debug(s"GET request to fetch employee with ID: $id")
              onComplete(service.getEmployeeById(id)) {
                case Success(Some(employee)) => complete(employee)
                case Success(None) => 
                  logger.warn(s"Employee not found with ID: $id")
                  complete(StatusCodes.NotFound -> ErrorResponse(s"Employee not found with ID: $id"))
                case Failure(ex) => 
                  logger.error(s"Error fetching employee $id: ${ex.getMessage}", ex)
                  complete(StatusCodes.InternalServerError -> ErrorResponse(s"Failed to get employee: ${ex.getMessage}"))
              }
            },
            put {
              entity(as[EmployeeRequest]) { request =>
                logger.debug(s"PUT request to update employee with ID: $id")
                onComplete(service.updateEmployee(id, request)) {
                  case Success(Some(employee)) => complete(employee)
                  case Success(None) => 
                    logger.warn(s"Employee not found with ID: $id")
                    complete(StatusCodes.NotFound -> ErrorResponse(s"Employee not found with ID: $id"))
                  case Failure(ex) => 
                    logger.error(s"Error updating employee $id: ${ex.getMessage}", ex)
                    complete(StatusCodes.InternalServerError -> ErrorResponse(s"Failed to update employee: ${ex.getMessage}"))
                }
              }
            },
            delete {
              logger.debug(s"DELETE request for employee with ID: $id")
              onComplete(service.deleteEmployee(id)) {
                case Success(true) => complete(StatusCodes.NoContent)
                case Success(false) => 
                  logger.warn(s"Employee not found with ID: $id")
                  complete(StatusCodes.NotFound -> ErrorResponse(s"Employee not found with ID: $id"))
                case Failure(ex) => 
                  logger.error(s"Error deleting employee $id: ${ex.getMessage}", ex)
                  complete(StatusCodes.InternalServerError -> ErrorResponse(s"Failed to delete employee: ${ex.getMessage}"))
              }
            }
          )
        }
      )
    }
  }
}
